### DISCUSSION TITLE
Better support for alternative solvers and solver options

### DISCUSSION THREAD

**Discussion by DillonJ** [spine-tools/SpineOpt.jl#322] [Ideas]
It would be nice to make it easier for users to use alternative solvers with SpineOpt.

Some of us have CPLEX working... and this apparently requires a few tricks. 
 - One needs to install the CPLEX.jl Julia plackage
 - You have to set an environment variable to point to the CPLEX dlls and build the package
 - You need to know how to set the `mip_solver` and `lp_solver` arguments in the run_spineopt call. I.e.:
   -  `mip_solver = optimizer_with_attributes(CPLEX.Optimizer, "CPX_PARAM_EPGAP" => 0.05)`
   - In the above call, you need to know the option names (GAMS for example has its own option names that it translates for each solver)
   - You need to include the CPLEX package in your SpineOpt tool code

At the very least, I think we need documentation on how to use other solvers. Ideally, we want the user to be able to specify the solver, perhaps in the model database or somewhere else in the toolbox

Thoughts @manuelma @mihlema @jkiviluo @Tasqu

**Comment by manuelma** [comment]
One thing I definitely don't like in our setup is that you need to have the JuMP function `optimizer_with_attributes` available *before* calling `run_spineopt`, which means you need to `import JuMP`. Ideally one should be able to just `import CPLEX`, pass `CPLEX.Optimizer` in the `lp_solver` argument, and the function internally would call `optimizer_with_attributes`.

But I guess we can consider doing *all* the 'tricks' listed in the description automatically? so the user basically doesn't need to learn or worry about anything. I like that approach. Of course there're things that can go wrong when you try to automatize steps that are intended for humans. There's a lot of cases you need to consider. Usually what works for somebody in one OS/platform doesn't work for somebody else in another OS/platform.

**Comment by mihlema** [comment]
> One needs to install the CPLEX.jl Julia plackage

I think that's fine , if that's the users responsibility to add the package. Usually the solver are quite easy to install, I wouldn't worry to much about that part

> You need to know how to set the mip_solver and lp_solver arguments in the run_spineopt call. I.e.:
> mip_solver = optimizer_with_attributes(CPLEX.Optimizer, "CPX_PARAM_EPGAP" => 0.05)
> In the above call, you need to know the option names (GAMS for example has its own option names that it translates for each solver)
> You need to include the CPLEX package in your SpineOpt tool code

Completely agree, this is very cumbersome at the moment. I'd like to see these optional arguments as parameters to the model object, that'd be neat - and it seems pretty straightforward to do.

However, I wouldn't know how we'd add the specified solver package with `run_spineopt` after calling `using_spinedb`.  Alternatively we could add a bunch of solvers to the dependencies of SpineOpt; then the user doesn't have to install them. This would probably add additional size and compiling time to SpineOpt, so this option doesn't seem ideal either.

**Comment by mihlema** [comment]
Links to #379
